### PR TITLE
Enable more configuration using the DTB

### PR DESCRIPTION
**Pull request by abejgonzalez**
Addresses parts of #1716 

Thanks @jerryz123 for the main work on this.

- A DTB that is passed in is used to set the number of harts, ISA, PMPs, MMU, and more of Spike.
  - Note: Device extensions that generate a DTS string do not affect the output of `--dump-dts` if the DTB is given.
  - Note: DTB provided memory regions are not used.
- Correctly catches interactive traps v.s. other traps in interactive mode.

### CODE REVIEW

**Comment by jerryz123** [comment]
Can you squash [751ab8f](https://github.com/riscv-software-src/riscv-isa-sim/pull/1721/commits/751ab8fad18cbcdfe9f2e7ceb618e9c60ce08ea7) into [433c16c](https://github.com/riscv-software-src/riscv-isa-sim/pull/1721/commits/433c16c542ad784a81640c8a9568f135ec0afcb2)? (But don't squash all the commits together).

File: riscv/dts.cc [hunk 1]
```diff
@@ -178,6 +181,88 @@ std::string dts_compile(const std::string& dts)
   return dtb.str();
 }
 
+std::string dtb_to_dts(const std::string& dtb)
```
Comment (jerryz123) [inline riscv/dts.cc]: Can this be DRY'd with `dts_to_dtb`? Perhaps both should just call a new `exec_dtc` with different flags.

**Comment by abejgonzalez** [comment]
> Can you squash [751ab8f](https://github.com/riscv-software-src/riscv-isa-sim/pull/1721/commits/751ab8fad18cbcdfe9f2e7ceb618e9c60ce08ea7) into [433c16c](https://github.com/riscv-software-src/riscv-isa-sim/pull/1721/commits/433c16c542ad784a81640c8a9568f135ec0afcb2)? (But don't squash all the commits together).

Done

File: riscv/dts.cc (hunk 1 above)
Comment (abejgonzalez) [reply riscv/dts.cc]: Done.

**Comment by aswaterman** [comment]
I don't expect to have time to review this one in the next few days, but @jerryz123 has my proxy.

**Comment by abejgonzalez** [comment]
> I don't expect to have time to review this one in the next few days, but @jerryz123 has my proxy.

No rush. Unfortunately, @jerryz123 wrote the main parts of this PR so I don't think it's fair to have him also review it :)

**Comment by aswaterman** [comment]
True.  Well, LMK when it's ready for review and I'll take a look when I have time.

**Comment by abejgonzalez** [comment]
> True. Well, LMK when it's ready for review and I'll take a look when I have time.

It's ready whenever you are. Thanks!

**Comment by abejgonzalez** [comment]
@aswaterman any update on this?

**Comment by abejgonzalez** [comment]
CI issues look unrelated to this PR (timing out on fetching openocd).

**Comment by abejgonzalez** [comment]
@jerryz123 Can you re-run the CI test that is failing?

**Comment by aswaterman** [comment]
I clicked the rerun button.

I hope to review the PR tomorrow.

**Comment by abejgonzalez** [comment]
> I clicked the rerun button.
> 
> I hope to review the PR tomorrow.

I think there is a problem with RISC-V OpenOCD. Looks like the https://git.savannah.nongnu.org/git/git2cl.git/ repo doesn't work anymore. 

In any case, thanks!

**Comment by aswaterman** [comment]
It’s probably a transient outage; let’s try again tomorrow.

**Comment by abejgonzalez** [comment]
> It’s probably a transient outage; let’s try again tomorrow.

CI passed so ready when you are. Thanks!

**Comment by aswaterman** [comment]
Hopefully tomorrow...

**Review by aswaterman — APPROVED** [review]
Thanks for organizing this PR such that it was easy to review.

There's one bit of refactoring I want to do, but I'll take care of it as my penance for taking so long.

**Comment by abejgonzalez** [comment]
Thanks!

**Pull request merged**